### PR TITLE
Remove unit dependences

### PR DESCRIPTION
**Pull request by eosti**
Currently, `worstcase` requires the use of `pint`. My desired workflow involves using `worstcase` with `handcalcs`, which supports the unit library `forallpeople` much better than it does `pint`, so my existing formulas and calculations all use `forallpeople` units. Since `worstcase` is, at its core, a computation library, I would like to use any (or none) unit library instead of being forced into using `pint` -- the computations in `worstcase` are unit agnostic, so the unit library is really more of a container for holding numeric data. 

This PR removes the dependency on `pint` and replaces it with duck-typing which can pull the value and units from an unknown datatype. This would allow `worstcase` to work with any of the common Python unit libraries, choosable by the end user. This change will also remove the exported `unit` object, since `worstcase` would no longer provide units itself. As such, this is a breaking change; users would need to configure the `UnitRegistry` themselves. 

I'm opening this as a draft for feedback before I finish all the implementation details: is this a desired improvement? Any additional functionality that would be useful? Is there anything I should know about `pint` that was found during development? I can confirm that the PR as it stands today works with both `pint` and `forallpeople` but I haven't spent a ton of time testing edge cases outside getting things working for my current analyses. 

Todos:
- add support for additional unit libraries (including unit tests)
- update examples in repo
- improve docs
- ???

### CODE REVIEW

**Comment by amosborne** [comment]
I think this is a pretty rad idea.

**Comment by eosti** [comment]
The only outstanding task is waiting for https://github.com/connorferster/forallpeople/pull/120 to get merged so that `pyproject.yml` doesn't have to point towards my personal fork with that commit applied. Aside from that, this PR is ready for review!

**Comment by amosborne** [comment]
Thanks Reid-- just a heads up that it might take me a week or two to dig into this.

**Comment by amosborne** [comment]
@eosti still working on this? just ping me when you're ready

**Comment by eosti** [comment]
@amosborne ready for review!

**Comment by amosborne** [comment]
***editorial comment -- claude outputs the following code review result; will post my own feedback separately***

## Review

The core duck-typed unit design works for the happy path in all five libraries, but I found one **silently-wrong-numbers bug** in the RSS path, several library-specific holes in the new unit validation, and a test parametrization flaw that is masking one of the bugs. All repros below were verified by execution against this branch.

---

### 🔴 High severity

#### 1. `root_sum_square` produces silently wrong bounds for mixed-but-compatible units
`worstcase/worstcase.py` (~L590)

`get_val()` strips raw magnitudes and L615-616 relabel them with `nom`'s units. On master this state was unreachable (strict `lb.u == nom.u == ub.u` assert), but the new `compare_units` checks *compatibility*, not equality, so mixed units now get through:

```python
A = param.byrange(4 * unit.V, 3999 * unit.mV, 5 * unit.V)
B = param.bytol(2 * unit.V, 0.1 * unit.V, False)
C = derive.byrss(A, B)(lambda a, b: a + b)
C()  # ub = 5999.0 V — correct answer is ~7.005 V
```

Pint evaluates `3999 mV + 2 V` in mV, `get_val()` returns the mV magnitude, and it gets relabeled as volts. No error, 1000x-wrong output. (`extreme_value` is **not** affected — it stays in quantity space throughout.) Fix: convert bounds to `nom`'s units before stripping, or normalize units in `Parameter.__init__`.

#### 2. forallpeople bypasses unit validation entirely
`worstcase/worstcase.py` (~L93)

The `val1 + val2` try-add probe silently passes when a forallpeople `Physical` is mixed with a bare number, because forallpeople coerces scalars (`2 + 1*si.V == 3 V`):

```python
param.bytol_abs(1 * si.A, 2)       # returns 1 A (nom), -1 A (lb), 3 A (ub) — should raise
param.byrange(1 * si.V, 0.5, 2)    # constructs unitful nom with unitless bounds — should raise
```

The `type(val1) is type(val2)` check rejects a dimensionless pint Quantity mixed with a plain number, which worked on master via the `nom * Unit([])` coercion:

```python
param.byrange(5 * unit.dimensionless, 4.9, 5.1)
# ValueError: Parameter bounds have inconsistent units.  (valid on master)
```

It also rejects Quantities from two different `UnitRegistry` instances. Root cause: the pint branch only exists because the try-add catches `ValueError`, while pint's `DimensionalityError` subclasses `TypeError` (MRO: `DimensionalityError → PintTypeError → TypeError`). Catching `(ValueError, TypeError)` in the shared try-add deletes the special case and fixes both problems.

---

### 🟡 Medium severity

#### 4. `bytol_pct` unit guard is wrong for 3 of 5 libraries
`worstcase/worstcase.py` (~L194)

The `get_units(pct) != 1` check:
- **python-quantities**: unitful pct passes the guard (`pq.A == 1` is `True`) → confusing downstream error `"Unable to convert between units of 'A*ohm' and 'ohm'"` instead of the intended message
- **forallpeople**: the `!=` comparison itself raises a subtraction error
- **unyt**: dimensionless quantities are wrongly *rejected* (`unyt.Unit('dimensionless') != 1`), while the equivalent pint/astropy/pq calls succeed

#### 5. `equivalent()` raises instead of returning `False` across unit libraries
`worstcase/worstcase.py` (~L110)

```python

#### 8. Test guards vacuously skip forallpeople and python-quantities
`tests/test_worstcase.py` (L63, L185)

`if amp_unit == 1: return` matches forallpeople and pq units too (`si.A == 1` and `pq.A == 1` are both `True`), so `test_param_invalid_input` and the cross-unit half of `test_param_equivalency` never run for those two libraries. **This is currently masking bug #2** — running the skipped body manually fails with `DID NOT RAISE ValueError`. Use the pytest param id or an `isinstance` check instead of `== 1`.

---

### 🟢 Low severity

#### 9. README examples no longer run
README code samples still use the removed `unit` object with no `UnitRegistry` construction shown (example.ipynb was updated; README was not). Pint isn't a runtime dependency anymore, so a user copying the README gets an `ImportError`/`NameError`.

#### 10. Ineffective regex in `test_import_units`
`tests/test_worstcase.py` (L256): `r"^worstcase no longer*"` quantifies the trailing `r`, so it only asserts the prefix `"worstcase no longe"`. Intended `r"^worstcase no longer"` (`match=` is already a search; no trailing wildcard needed).

---

### Cleanup notes (non-blocking)

- `root_sum_square` calls `get_val(nom)` 6× and `get_units(nom)` 2× — hoist into locals; same for the double `get_units(nom)` in `extreme_value`'s awkwardly-wrapped init
- `bytol_pct` docstring documents `tol` but the parameter is `pct`; `bytol_abs` shadows the builtin `abs`
- Deleting `test_derive_byrss_warnasymmetric` leaves the asymmetry `UserWarning` path with zero coverage — notable since bug #1 corrupts exactly that metric
- Library detection is spread across four independent `hasattr` probes (`get_units`, `get_val`, `compare_units`, `formatter_string`); a single per-library adapter would keep them from drifting

**Common thread for #1–#4:** master enforced strict unit *equality* via pint; this PR relaxes it to duck-typed *compatibility*, and each library's quirks leak through a different hole. The deepest fix is one shared try-add probe catching `(ValueError, TypeError)` plus normalizing bounds to `nom`'s units at construction.

File: worstcase/worstcase.py [hunk 1]
```diff
@@ -39,34 +121,106 @@ class Parameter(AbstractParameter):
...
+            nom: nominal value
+            lb: lower bound on the nominal value
+            ub: upper bound on the nominal value
+            tag: name of the parameter
+            sigfig: number of significant figures to print
+        """
         return Parameter(nom, lb, ub, tag, sigfig)
 
     @staticmethod
     def bytol(nom, tol, rel, tag="", sigfig=4):
```
Comment (amosborne) [inline worstcase/worstcase.py:132]: I'm not a huge fan of breaking up the `bytol` function since it feels like it adds code unnecessarily but it's not a hill I'm willing to die on. If so, `bytol_pct` should either assume that the number is already in percentage form (100=100%) or it should be called `bytol_rel` (for relative tolerance).

File: worstcase/worstcase.py [hunk 2]
```diff
@@ -39,34 +121,106 @@ class Parameter(AbstractParameter):
     def __init__(
         self, nom, lb, ub, tag, sigfig, derivation=Derivation(nom={}, lb={}, ub={})
```
Comment (amosborne) [inline worstcase/worstcase.py:81]: Looking at this signature now I am not a huge fan of having `sigfig` as an argument since it is purely intended to fine tune the display. We should remove it as an argument; parameters should "just work" with the standard python format-spec strings.

File: worstcase/worstcase.py [hunk 3]
```diff
@@ -6,31 +6,113 @@
...
+            # astropy Quantity instance
+            return value.unit
+        if hasattr(value, "dimensions"):
+            # forallpeople Physical instance
+            return value.split()[1]
+        if hasattr(value, "dimensionality"):
+            # python-quantities Quantity instance
+            return value.dimensionality
+        # Unknown type, return unitless
+        return 1
```
Comment (amosborne) [inline worstcase/worstcase.py]: There is an idiomatic way to do this that is far more robust than checking for attributes and then assuming which library the unit is from. We should be able to write an "adapter" for each unit library that explicitly checks which package the unit comes from and constructs a common set of static methods. Claude-generated example code below:  

```python
class ScalarAdapter:
    """Default: plain numbers, or anything we don't recognize."""
    magnitude = staticmethod(lambda q: q)
    units = staticmethod(lambda q: 1)
    def compatible(self, a, b): ...
    def format(self, q, sigfig): return format(q, f"0.{sigfig}g")

class PintAdapter(ScalarAdapter):
    magnitude = staticmethod(attrgetter("magnitude"))
    units = staticmethod(attrgetter("units"))
    def format(self, q, sigfig): return format(q, f"0.{sigfig}G#~P")

class AstropyAdapter(ScalarAdapter):
    magnitude = staticmethod(attrgetter("value"))
    units = staticmethod(attrgetter("unit"))
...

_ADAPTERS = {"pint": PintAdapter(), "astropy": AstropyAdapter(),
             "unyt": UnytAdapter(), "quantities": QuantitiesAdapter(),
             "forallpeople": ForallpeopleAdapter()}

@functools.lru_cache(maxsize=None)
def adapter_for(cls):
    return _ADAPTERS.get(cls.__module__.split(".")[0], ScalarAdapter())
```

File: worstcase/worstcase.py [hunk 4]
```diff
@@ -110,15 +264,19 @@ def apply(self, funcname, *args, **kwargs):
...
+        """Pint-only method, left for backwards compatiblity"""
         return self.apply("ito_base_units")
 
     def ito_reduced_units(self):
+        """Pint-only method, left for backwards compatiblity"""
         return self.apply("ito_reduced_units")
 
     def ito_root_units(self):
+        """Pint-only method, left for backwards compatiblity"""
         return self.apply("ito_root_units")
```
Comment (amosborne) [inline worstcase/worstcase.py]: I'm not concerned about backwards compatibility, we will release this as version 1.0.0. Pint-only methods can be removed. If we want to provide functionality such as is, it will need to be implemented across all unit-library adapters. I don't think I ever use these methods in practice.

File: worstcase/worstcase.py (hunk 2 above)
Comment (amosborne) [reply worstcase/worstcase.py:81]: I would be also pretty interested in removing the `tag` argument an instead inferring the parameter "name" using the package `varname`. This package inspects the AST to figure out the what the caller uses as the name for the object we are creating. As long as we don't use the name as load-bearing in the code, it should make the api a little more ergonomic.

Comment (eosti) [reply worstcase/worstcase.py:81]: Both of these improvements sound great, but to keep this PR remotely on-topic I've created issues (#5 and #6) to address these after this PR has gotten merged.

File: worstcase/worstcase.py (hunk 1 above)
Comment (eosti) [reply worstcase/worstcase.py:132]: I've been pretty convinced away from flag arguments thanks to https://martinfowler.com/bliki/FlagArgument.html. Looking through analysis I've written using the existing API, it's hard to quickly remember what omitting the `rel` flag means, or why there's a `True` as the last argument to what feels like should be a purely numeric function. You could manage it using a forced keyword-only argument (https://ivergara.github.io/boolean-arguments-to-functions-in-python.html) but that still allows you to omit the parameter entirely which hurts readability. 

I chose to leave the original `bytol` function so that old code won't break, and while it adds additional lines it leaves the logic in the `bytol_*` functions so there's no logic duplicated. 

I agree that it should be called `bytol_rel` to match the original function call; I'll make that change.

File: worstcase/worstcase.py (hunk 4 above)
Comment (eosti) [reply worstcase/worstcase.py]: Neat, I'll remove these.

**Comment by eosti** [comment]
Thanks for the review, I've addressed the super easy stuff (and some embarrassing typos that snuck in) and will look into the other items in the next week or two. I really like the adapter idea, although I'll need to learn more about functools to properly wrap my head around it.

File: worstcase/worstcase.py (hunk 1 above)
Comment (amosborne) [reply worstcase/worstcase.py:132]: Hm that is a reasonable argument; I agree. Nah forced keyword arguments are an anti-pattern in my opinion.

**Comment by amosborne** [comment]
Sure-- I think the `functools.lru_cache` is just some extra sauce so the adapter call doesn't get computed a million times.

**Comment by eosti** [comment]
I added the unit adapter functionality as a separate file along with a small test suite. This seems to be much more pythonic, and the `lru_cache` was in fact not a big deal once I read into it 😄. This is now ready for review again. 

For the Claude review:

1. Fixed with the unit adapter refactor. Bounds are converted to `nom`'s units during initialization (i.e. `param.bytol`, etc) to avoid these scaling issues. This was a good catch. 
2. Added stricter typechecking around value types needing to match. 
3. 👻 
4. There's more tests to make sure that weird values get caught by `bytol_rel`, and there's now a check for dimensionless instead of just non-existent units. 
5. Wasn't able to reproduce this, although there are now checks to verify that the units match first which should cut down on potential mismatches. 
6. 👻 
7. 👻 
8. Good catch, now checks against the unit ID. 
9. Fixed
10. Fixed

There's some fixups for ease of review; once approved I'll squash stuff together prior to the merge.

File: worstcase/worstcase.py (hunk 3 above)
Comment (eosti) [reply worstcase/worstcase.py]: Added as `unit_adapter.py`.
